controllers/service_labels/abstract.go

```diff
@@ -0,0 +1,142 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service_labels
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/clastix/capsule/api/v1alpha1"
+)
+
+type abstractServiceLabelsReconciler struct {
+	obj    client.Object
+	client client.Client
+	log    logr.Logger
+	scheme *runtime.Scheme
+}
+
+func (r *abstractServiceLabelsReconciler) InjectClient(c client.Client) error {
+	r.client = c
+	return nil
+}
+
+func (r *abstractServiceLabelsReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	tenant, err := r.getTenant(ctx, request.NamespacedName, r.client)
+	if err != nil {
+		switch err.(type) {
+		case *NonTenantObject, *NoServicesMetadata:
+			return reconcile.Result{}, nil
+		default:
+			r.log.Error(err, fmt.Sprintf("Cannot sync %t labels", r.obj))
+			return reconcile.Result{}, err
+		}
+	}
+
+	err = r.client.Get(ctx, request.NamespacedName, r.obj)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, r.client, r.obj, func() (err error) {
+		r.obj.SetLabels(r.sync(r.obj.GetLabels(), tenant.Spec.ServicesMetadata.AdditionalLabels))
+		r.obj.SetAnnotations(r.sync(r.obj.GetAnnotations(), tenant.Spec.ServicesMetadata.AdditionalAnnotations))
+		return nil
+	})
+
+	return reconcile.Result{}, err
+}
+
+func (r *abstractServiceLabelsReconciler) getTenant(ctx context.Context, namespacedName types.NamespacedName, client client.Client) (*v1alpha1.Tenant, error) {
+	ns := &corev1.Namespace{}
+	tenant := &v1alpha1.Tenant{}
+
+	if err := client.Get(ctx, types.NamespacedName{Name: namespacedName.Namespace}, ns); err != nil {
+		return nil, err
+	}
+
+	capsuleLabel, _ := v1alpha1.GetTypeLabel(&v1alpha1.Tenant{})
+	if _, ok := ns.GetLabels()[capsuleLabel]; !ok {
+		return nil, NewNonTenantObject(namespacedName.Name)
+	}
+
+	if err := client.Get(ctx, types.NamespacedName{Name: ns.Labels[capsuleLabel]}, tenant); err != nil {
+		return nil, err
+	}
+
+	if tenant.Spec.ServicesMetadata.AdditionalLabels == nil && tenant.Spec.ServicesMetadata.AdditionalAnnotations == nil {
+		return nil, NewNoServicesMetadata(namespacedName.Name)
+	}
+
+	return tenant, nil
+}
+
+func (r *abstractServiceLabelsReconciler) sync(available map[string]string, tenantSpec map[string]string) map[string]string {
+	if tenantSpec != nil {
+		if available == nil {
+			available = make(map[string]string)
```

Comment:
The linter here is right: why not directly

```go
available = tenantSpec
```

I don't get the point of initializing a variable and then overwrite it.